fs_impl.go

```diff
@@ -0,0 +1,353 @@
+package main
+
+// the directory tree is modeled using the Records structs
+// they are reference counted and thus are forbidden to form cycles
+
+// Records are saved as $fs_root/$uuid
+
+// Records contain sections saved as $fs_root/$uuid.$section The file payload
+// is saved in the 'data' section. metadata is in 'meta'. hooks can create own
+// sections
+
+import (
+	"encoding/json"
+	"errors"
+	"github.com/google/uuid"
+	"net/http"
+	"os"
+	"path/filepath"
+	"regexp"
+	"strings"
+)
+
+const (
+	sectionPattern      = `[a-zA-Z0-9_-]+`
+	uuidPattern         = `[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}`
+	fileInFsRootPattern = uuidPattern + `(\.` + sectionPattern + `)?`
+
+	onlyUuidPattern         = `^` + uuidPattern + `$`
+	onlyFileInFsRootPattern = `^` + fileInFsRootPattern + `$`
+	onlySectionPattern      = `^` + sectionPattern + `$`
+)
+
+type Record struct {
+	Children []uuid.UUID `json:"children,omitempty"`
+	IsDir    bool        `json:"is_dir"`
+	Name     string      `json:"name"`
+	refs     uint        `json:"-"`
+}
+
+type Fs struct {
+	records  map[uuid.UUID]*Record
+	root     uuid.UUID
+	basePath string
+}
+
+func (fs *Fs) path(p string) string {
+	// TODO sanitize paths
+	return filepath.Join(fs.basePath, p)
+}
+
+func (fs *Fs) writeToDisk(id uuid.UUID, r *Record) error {
```

Comment:
Why is this named `writeToDisk`? Since this implementation of fs uses the disk to store data, this should be implicit.

Reply:
This internal helper function writes the record to disk so i named it writeToDisk. feel free to rename it. i dont have strong feelings about it